src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -73,17 +74,51 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
 
             var targetStatement = (StatementSyntax)node;
             var leftmostCondition = GetLeftmostCondition(targetStatement);
-            if (!leftmostCondition.IsKind(SyntaxKind.NotEqualsExpression, out BinaryExpressionSyntax notEquals))
+            if (leftmostCondition == null)
             {
                 return;
             }
 
-            var operand = GetNullCheckOperand(notEquals.Left, notEquals.Right)?.WalkDownParentheses();
+            bool isNegativeNullCheck;
+            switch (leftmostCondition.Kind())
+            {
+                case SyntaxKind.NotEqualsExpression:
+                    isNegativeNullCheck = false;
+                    break;
+                case SyntaxKind.EqualsExpression:
+                    isNegativeNullCheck = true;
+                    break;
+                default:
+                    return;
+            }
+
+            // The pattern variable is only definitly assigned if the pattern succeeded,
```

Comment:
:bulb: definit**e**ly

:bulb: I find this block pleasantly readable, but it may be because I have context from seeing the recent bug. Consider expanding this comment by adding the following:

```csharp
// For example:
//
// if ((x = o as string) != null || SomeExpression) {
//   // x would not be definitely assigned here if pattern matching were used
// }
```